File: 2026-06-16-conflux/conflux/solver.py
"""A from-scratch CDCL SAT solver.

Implements the modern conflict-driven clause-learning algorithm:
  * two-watched-literal unit propagation
  * VSIDS branching heuristic (activity bumping + decay)
  * 1-UIP conflict analysis -> learned clause
  * recursive learned-clause minimization
  * non-chronological backjumping
  * Luby-sequence restarts
  * phase saving

The solver can optionally record a DRAT-style proof (the sequence of learned
clauses) so that an UNSAT result is independently checkable (see proof.py).
"""
from __future__ import annotations

import heapq
from dataclasses import dataclass, field
from typing import List, Optional, Dict, Tuple

from .cnf import CNF

SAT = "SAT"
UNSAT = "UNSAT"
UNKNOWN = "UNKNOWN"


def luby(i: int) -> int:
    """The Luby sequence (1-indexed): 1,1,2,1,1,2,4,1,1,2,1,1,2,4,8,..."""
    # Knuth's closed form
    k = 1
    while True:
        if i == (1 << k) - 1:
            return 1 << (k - 1)
        if (1 << (k - 1)) <= i < (1 << k) - 1:
            return luby(i - (1 << (k - 1)) + 1)
        k += 1


@dataclass
class Stats:
    decisions: int = 0
    propagations: int = 0
    conflicts: int = 0
    learned: int = 0
    restarts: int = 0
    max_decision_level: int = 0
    minimized_lits: int = 0


class Solver:
    def __init__(self, num_vars: int = 0, record_proof: bool = False,
                 var_decay: float = 0.95, restart_base: int = 100):
        self.num_vars = num_vars
        # value[v] in {None, True, False}; index 0 unused
        self.val: List[Optional[bool]] = [None] * (num_vars + 1)
        self.level: List[int] = [-1] * (num_vars + 1)
        self.reason: List[Optional[int]] = [None] * (num_vars + 1)
        self.phase: List[bool] = [False] * (num_vars + 1)

        self.clauses: List[List[int]] = []      # both original and learned
        self.is_learnt: List[bool] = []
        self.num_original = 0
        self.watches: Dict[int, List[int]] = {}  # literal -> clause refs

        self.trail: List[int] = []
        self.trail_lim: List[int] = []
        self.qhead = 0

        # VSIDS
        self.activity: List[float] = [0.0] * (num_vars + 1)
        self.var_inc = 1.0
        self.var_decay = var_decay
        self.heap: List[Tuple[float, int]] = []  # (-activity, var), lazy

        self.ok = True
        self.stats = Stats()
        self.restart_base = restart_base

        # proof recording (DRAT additions; deletions omitted -> still valid RUP)
        self.record_proof = record_proof
        self.proof: List[List[int]] = []

        # optional structured tracer (set externally)
        self.tracer = None

    # ---- helpers ---------------------------------------------------------
    def _ensure_var(self, v: int) -> None:
        if v <= self.num_vars:
            return
        extra = v - self.num_vars
        self.val.extend([None] * extra)
        self.level.extend([-1] * extra)
        self.reason.extend([None] * extra)
        self.phase.extend([False] * extra)
        self.activity.extend([0.0] * extra)
        self.num_vars = v

    def _w(self, lit: int) -> List[int]:
        lst = self.watches.get(lit)
        if lst is None:
            lst = []
            self.watches[lit] = lst
        return lst

    def lit_val(self, lit: int) -> Optional[bool]:
        v = self.val[abs(lit)]
        if v is None:
            return None
        return v if lit > 0 else (not v)

    def lit_true(self, lit: int) -> bool:
        return self.lit_val(lit) is True

    def lit_false(self, lit: int) -> bool:
        return self.lit_val(lit) is False

    def decision_level(self) -> int:
        return len(self.trail_lim)

    # ---- clause database -------------------------------------------------
    def add_clause(self, lits) -> bool:
        """Add an *original* clause. Returns False if it makes the formula UNSAT."""
        if not self.ok:
            return False
        seen = set()
        cl: List[int] = []
        for l in lits:
            l = int(l)
            if l == 0:
                raise ValueError("0 is not a literal")
            self._ensure_var(abs(l))
            if -l in seen:
                return True  # tautology: a v ~a v ... is always satisfied
            if l in seen:
                continue
            seen.add(l)
            cl.append(l)
        if len(cl) == 0:
            self.ok = False
            return False
        if len(cl) == 1:
            return self._enqueue_initial(cl[0])
        cref = len(self.clauses)
        self.clauses.append(cl)
        self.is_learnt.append(False)
        self.num_original += 1
        self._w(cl[0]).append(cref)
        self._w(cl[1]).append(cref)
        return True

    def _enqueue_initial(self, lit: int) -> bool:
        v = self.lit_val(lit)
        if v is True:
            return True
        if v is False:
            self.ok = False
            return False
        self._enqueue(lit, None)
        return True

    def _add_learnt(self, lits: List[int]) -> int:
        cref = len(self.clauses)
        self.clauses.append(lits)
        self.is_learnt.append(True)
        if len(lits) >= 2:
            self._w(lits[0]).append(cref)
            self._w(lits[1]).append(cref)
        return cref

    # ---- assignment ------------------------------------------------------
    def _enqueue(self, lit: int, reason: Optional[int]) -> None:
        v = abs(lit)
        self.val[v] = (lit > 0)
        self.level[v] = self.decision_level()
        self.reason[v] = reason
        self.trail.append(lit)

    def _new_decision_level(self) -> None:
        self.trail_lim.append(len(self.trail))

    def _backtrack_to(self, level: int) -> None:
        if self.decision_level() <= level:
            return
        target = self.trail_lim[level]
        for k in range(len(self.trail) - 1, target - 1, -1):
            lit = self.trail[k]
            v = abs(lit)
            self.phase[v] = self.val[v]  # phase saving
            self.val[v] = None
            self.reason[v] = None
            self.level[v] = -1
            heapq.heappush(self.heap, (-self.activity[v], v))
        del self.trail[target:]
        del self.trail_lim[level:]
        self.qhead = min(self.qhead, len(self.trail))

    # ---- VSIDS -----------------------------------------------------------
    def _bump(self, v: int) -> None:
        self.activity[v] += self.var_inc
        if self.activity[v] > 1e100:
            self._rescale()
        heapq.heappush(self.heap, (-self.activity[v], v))

    def _rescale(self) -> None:
        for v in range(1, self.num_vars + 1):
            self.activity[v] *= 1e-100
        self.var_inc *= 1e-100
        # heap absolute values now stale -> rebuild from unassigned vars
        self.heap = [(-self.activity[v], v)
                     for v in range(1, self.num_vars + 1) if self.val[v] is None]
        heapq.heapify(self.heap)

    def _decay(self) -> None:
        self.var_inc /= self.var_decay

    def _pick_branch_var(self) -> Optional[int]:
        while self.heap:
            negact, v = heapq.heappop(self.heap)
            if self.val[v] is None:
                return v
        # heap exhausted (can happen after rescale/rebuild) — linear fallback
        best, best_act = None, -1.0
        for v in range(1, self.num_vars + 1):
            if self.val[v] is None and self.activity[v] >= best_act:
                best, best_act = v, self.activity[v]
        return best

    # ---- propagation -----------------------------------------------------
    def propagate(self) -> Optional[int]:
        while self.qhead < len(self.trail):
            p = self.trail[self.qhead]
            self.qhead += 1
            self.stats.propagations += 1
            neg = -p  # literals that just became FALSE
            ws = self.watches.get(neg, [])
            new_ws: List[int] = []
            i = 0
            conflict = None
            n = len(ws)
            while i < n:
                cref = ws[i]
                i += 1
                c = self.clauses[cref]
                # ensure the false literal sits at c[1]
                if c[0] == neg:
                    c[0], c[1] = c[1], c[0]
                first = c[0]
                if first != neg and self.lit_true(first):
                    new_ws.append(cref)  # clause already satisfied
                    continue
                # find a new, non-false literal to watch
                found = False
                for k in range(2, len(c)):
                    if not self.lit_false(c[k]):
                        c[1], c[k] = c[k], c[1]
                        self._w(c[1]).append(cref)
                        found = True
                        break
                if found:
                    continue
                # no replacement: keep watching neg
                new_ws.append(cref)
                if self.lit_false(first):
                    conflict = cref
                    while i < n:  # preserve the rest of the watch list
                        new_ws.append(ws[i])
                        i += 1
                    break
                else:
                    self._enqueue(first, cref)
            self.watches[neg] = new_ws
            if conflict is not None:
                self.qhead = len(self.trail)
                return conflict
        return None

    # ---- conflict analysis (1-UIP) --------------------------------------
    def analyze(self, confl: int) -> Tuple[List[int], int]:
        learnt: List[int] = [0]  # placeholder for the asserting literal
        seen = [False] * (self.num_vars + 1)
        path_count = 0
        p = 0
        index = len(self.trail) - 1
        cur_level = self.decision_level()
        confl_lits = self.clauses[confl]

        while True:
            start = 0 if p == 0 else 1
            for j in range(start, len(confl_lits)):
                q = confl_lits[j]
                v = abs(q)
                if not seen[v] and self.level[v] > 0:
                    seen[v] = True
                    self._bump(v)
                    if self.level[v] >= cur_level:
                        path_count += 1
                    else:
                        learnt.append(q)
            # find the next literal on the trail that we've seen
            while not seen[abs(self.trail[index])]:
                index -= 1
            p = self.trail[index]
            seen[abs(p)] = False
            index -= 1
            path_count -= 1
            if path_count <= 0:
                break
            r = self.reason[abs(p)]
            assert r is not None, "implied literal must have a reason"
            confl_lits = self.clauses[r]

        learnt[0] = -p  # asserting literal = negation of the UIP

        # ---- recursive minimization -------------------------------------
        before = len(learnt)
        learnt = self._minimize(learnt, seen)
        self.stats.minimized_lits += before - len(learnt)

        # ---- compute backjump level -------------------------------------
        if len(learnt) == 1:
            bt = 0
        else:
            # second-highest level literal goes to position 1
            max_i = 1
            for i in range(2, len(learnt)):
                if self.level[abs(learnt[i])] > self.level[abs(learnt[max_i])]:
                    max_i = i
            learnt[1], learnt[max_i] = learnt[max_i], learnt[1]
            bt = self.level[abs(learnt[1])]
        return learnt, bt

    def _minimize(self, learnt: List[int], seen: List[bool]) -> List[int]:
        """Recursive self-subsumption: drop literals whose negation is implied by
        the other learnt literals (i.e. redundant given their reasons)."""
        # `seen` currently marks all vars in `learnt` (analyze left them set
        # except the UIP which it cleared). Rebuild reliably:
        mark = set(abs(l) for l in learnt)
        out = [learnt[0]]
        for l in learnt[1:]:
            r = self.reason[abs(l)]
            if r is None:
                out.append(l)  # decision literal — must stay
            elif not self._redundant(l, mark):
                out.append(l)
        return out

    def _redundant(self, lit: int, mark: set, depth: int = 0) -> bool:
        r = self.reason[abs(lit)]
        if r is None:
            return False
        if depth > 200:  # guard against pathological recursion
            return False
        for q in self.clauses[r]:
            if abs(q) == abs(lit):
                continue
            v = abs(q)
            if self.level[v] == 0:
                continue  # level-0 literals are globally fixed
            if v in mark:
                continue
            if self.reason[v] is None:
                return False  # depends on a decision not in the clause
            if not self._redundant(q, mark, depth + 1):
                return False
        return True

    # ---- main search -----------------------------------------------------
    def solve(self, max_conflicts: Optional[int] = None) -> str:
        if not self.ok:
            if self.record_proof:
                self.proof.append([])  # empty clause
            return UNSAT
        # init activity heap
        if not self.heap:
            self.heap = [(-self.activity[v], v) for v in range(1, self.num_vars + 1)]
            heapq.heapify(self.heap)

        # propagate any initial unit clauses
        if self.propagate() is not None:
            self.ok = False
            if self.record_proof:
                self.proof.append([])
            return UNSAT

        restart_no = 0
        conflicts_until_restart = self.restart_base * luby(restart_no + 1)
        conflicts_this_run = 0

        while True:
            confl = self.propagate()
            if confl is not None:
                self.stats.conflicts += 1
                conflicts_this_run += 1
                if max_conflicts is not None and self.stats.conflicts > max_conflicts:
                    return UNKNOWN
                if self.tracer:
                    self.tracer.on_conflict(self, confl)
                if self.decision_level() == 0:
                    if self.record_proof:
                        self.proof.append([])
                    return UNSAT
                learnt, bt = self.analyze(confl)
                self.stats.learned += 1
                if self.record_proof:
                    self.proof.append(list(learnt))
                if self.tracer:
                    self.tracer.on_learnt(self, learnt, bt)
                self._backtrack_to(bt)
                if len(learnt) == 1:
                    self._enqueue(learnt[0], None)
                else:
                    cref = self._add_learnt(learnt)
                    self._enqueue(learnt[0], cref)
                self._decay()
            else:
                if self.decision_level() > self.stats.max_decision_level:
                    self.stats.max_decision_level = self.decision_level()
                if conflicts_this_run >= conflicts_until_restart:
                    self.stats.restarts += 1
                    restart_no += 1
                    conflicts_this_run = 0
                    conflicts_until_restart = self.restart_base * luby(restart_no + 1)
                    self._backtrack_to(0)
                    continue
                v = self._pick_branch_var()
                if v is None:
                    return SAT  # all variables assigned, no conflict
                self.stats.decisions += 1
                self._new_decision_level()
                lit = v if self.phase[v] else -v
                if self.tracer:
                    self.tracer.on_decision(self, lit)
                self._enqueue(lit, None)

    def model(self) -> Dict[int, bool]:
        return {v: bool(self.val[v]) if self.val[v] is not None else False
                for v in range(1, self.num_vars + 1)}


def solve_cnf(cnf: CNF, record_proof: bool = False, **kw):
    """Convenience: build a solver from a CNF and solve.

    Returns (result, model_or_None, solver).
    """
    s = Solver(num_vars=cnf.num_vars, record_proof=record_proof, **kw)
    for cl in cnf.clauses:
        if not s.add_clause(cl):
            break
    res = s.solve()
    mdl = s.model() if res == SAT else None
    return res, mdl, s

File: 2026-06-16-conflux/web/solver.js
// Conflux CDCL solver — JavaScript mirror of conflux/solver.py.
// Faithful port: two-watched literals, VSIDS, 1-UIP analysis with recursive
// minimization, non-chronological backjumping, Luby restarts, phase saving.
// Optionally records a step trace for the visualizer.
//
// Runs both in the browser (attaches to window.Conflux) and under Node
// (module.exports) so the same code is checked for Python<->JS parity.
(function (root) {
  "use strict";

  function luby(i) {
    let k = 1;
    while (true) {
      if (i === (1 << k) - 1) return 1 << (k - 1);
      if ((1 << (k - 1)) <= i && i < (1 << k) - 1) return luby(i - (1 << (k - 1)) + 1);
      k += 1;
    }
  }

  // Minimal binary heap of {act, v} keyed by (-act, v) to mirror Python heapq
  // on tuples (-activity, var): higher activity first, lower var breaks ties.
  class Heap {
    constructor() { this.a = []; }
    get size() { return this.a.length; }
    _less(x, y) { // returns true if x should be above y
      if (x.act !== y.act) return x.act > y.act;
      return x.v < y.v;
    }
    push(item) {
      const a = this.a; a.push(item);
      let i = a.length - 1;
      while (i > 0) {
        const p = (i - 1) >> 1;
        if (this._less(a[i], a[p])) { [a[i], a[p]] = [a[p], a[i]]; i = p; }
        else break;
      }
    }
    pop() {
      const a = this.a; if (a.length === 0) return null;
      const top = a[0], last = a.pop();
      if (a.length) {
        a[0] = last; let i = 0;
        while (true) {
          let l = 2 * i + 1, r = 2 * i + 2, m = i;
          if (l < a.length && this._less(a[l], a[m])) m = l;
          if (r < a.length && this._less(a[r], a[m])) m = r;
          if (m === i) break;
          [a[i], a[m]] = [a[m], a[i]]; i = m;
        }
      }
      return top;
    }
  }

  class Solver {
    constructor(numVars, opts) {
      opts = opts || {};
      this.numVars = numVars;
      this.val = new Array(numVars + 1).fill(null);
      this.level = new Array(numVars + 1).fill(-1);
      this.reason = new Array(numVars + 1).fill(null);
      this.phase = new Array(numVars + 1).fill(false);

      this.clauses = [];
      this.isLearnt = [];
      this.numOriginal = 0;
      this.watches = new Map();

      this.trail = [];
      this.trailLim = [];
      this.qhead = 0;

      this.activity = new Array(numVars + 1).fill(0);
      this.varInc = 1.0;
      this.varDecay = opts.varDecay || 0.95;
      this.heap = new Heap();

      this.ok = true;
      this.restartBase = opts.restartBase || 100;
      this.stats = { decisions: 0, propagations: 0, conflicts: 0, learned: 0,
                     restarts: 0, maxLevel: 0 };

      this.recordTrace = !!opts.recordTrace;
      this.trace = [];
      this.maxTrace = opts.maxTrace || 200000;
    }

    _emit(ev) { if (this.recordTrace && this.trace.length < this.maxTrace) this.trace.push(ev); }

    _w(lit) {
      let l = this.watches.get(lit);
      if (!l) { l = []; this.watches.set(lit, l); }
      return l;
    }
    litVal(lit) {
      const v = this.val[Math.abs(lit)];
      if (v === null) return null;
      return lit > 0 ? v : !v;
    }
    litTrue(lit) { return this.litVal(lit) === true; }
    litFalse(lit) { return this.litVal(lit) === false; }
    decisionLevel() { return this.trailLim.length; }

    addClause(lits) {
      if (!this.ok) return false;
      const seen = new Set(), cl = [];
      for (let l of lits) {
        l = l | 0;
        if (l === 0) throw new Error("0 is not a literal");
        if (seen.has(-l)) return true; // tautology
        if (seen.has(l)) continue;
        seen.add(l); cl.push(l);
      }
      if (cl.length === 0) { this.ok = false; return false; }
      if (cl.length === 1) return this._enqueueInitial(cl[0]);
      const cref = this.clauses.length;
      this.clauses.push(cl); this.isLearnt.push(false); this.numOriginal++;
      this._w(cl[0]).push(cref); this._w(cl[1]).push(cref);
      return true;
    }
    _enqueueInitial(lit) {
      const v = this.litVal(lit);
      if (v === true) return true;
      if (v === false) { this.ok = false; return false; }
      this._enqueue(lit, null); return true;
    }
    _addLearnt(lits) {
      const cref = this.clauses.length;
      this.clauses.push(lits); this.isLearnt.push(true);
      if (lits.length >= 2) { this._w(lits[0]).push(cref); this._w(lits[1]).push(cref); }
      return cref;
    }
    _enqueue(lit, reason) {
      const v = Math.abs(lit);
      this.val[v] = lit > 0;
      this.level[v] = this.decisionLevel();
      this.reason[v] = reason;
      this.trail.push(lit);
    }
    _newDecisionLevel() { this.trailLim.push(this.trail.length); }

    _backtrackTo(level) {
      if (this.decisionLevel() <= level) return;
      const target = this.trailLim[level];
      for (let k = this.trail.length - 1; k >= target; k--) {
        const v = Math.abs(this.trail[k]);
        this.phase[v] = this.val[v];
        this.val[v] = null; this.reason[v] = null; this.level[v] = -1;
        this.heap.push({ act: this.activity[v], v: v });
      }
      this.trail.length = target;
      this.trailLim.length = level;
      if (this.qhead > this.trail.length) this.qhead = this.trail.length;
    }

    _bump(v) {
      this.activity[v] += this.varInc;
      if (this.activity[v] > 1e100) this._rescale();
      this.heap.push({ act: this.activity[v], v: v });
    }
    _rescale() {
      for (let v = 1; v <= this.numVars; v++) this.activity[v] *= 1e-100;
      this.varInc *= 1e-100;
      this.heap = new Heap();
      for (let v = 1; v <= this.numVars; v++)
        if (this.val[v] === null) this.heap.push({ act: this.activity[v], v: v });
    }
    _decay() { this.varInc /= this.varDecay; }

    _pickBranchVar() {
      while (this.heap.size) {
        const it = this.heap.pop();
        if (this.val[it.v] === null) return it.v;
      }
      let best = null, bestAct = -1;
      for (let v = 1; v <= this.numVars; v++)
        if (this.val[v] === null && this.activity[v] >= bestAct) { best = v; bestAct = this.activity[v]; }
      return best;
    }

    propagate() {
      while (this.qhead < this.trail.length) {
        const p = this.trail[this.qhead++];
        this.stats.propagations++;
        const neg = -p;
        const ws = this.watches.get(neg) || [];
        const newWs = [];
        let conflict = null, i = 0; const n = ws.length;
        while (i < n) {
          const cref = ws[i++]; const c = this.clauses[cref];
          if (c[0] === neg) { const t = c[0]; c[0] = c[1]; c[1] = t; }
          const first = c[0];
          if (first !== neg && this.litTrue(first)) { newWs.push(cref); continue; }
          let found = false;
          for (let k = 2; k < c.length; k++) {
            if (!this.litFalse(c[k])) { const t = c[1]; c[1] = c[k]; c[k] = t; this._w(c[1]).push(cref); found = true; break; }
          }
          if (found) continue;
          newWs.push(cref);
          if (this.litFalse(first)) {
            conflict = cref;
            while (i < n) newWs.push(ws[i++]);
            break;
          } else {
            this._enqueue(first, cref);
            this._emit({ t: "imply", lit: first, reason: c.slice(), level: this.decisionLevel() });
          }
        }
        this.watches.set(neg, newWs);
        if (conflict !== null) { this.qhead = this.trail.length; return conflict; }
      }
      return null;
    }

    analyze(confl) {
      const learnt = [0];
      const seen = new Array(this.numVars + 1).fill(false);
      let pathCount = 0, p = 0, index = this.trail.length - 1;
      const curLevel = this.decisionLevel();
      let lits = this.clauses[confl];
      while (true) {
        const start = p === 0 ? 0 : 1;
        for (let j = start; j < lits.length; j++) {
          const q = lits[j], v = Math.abs(q);
          if (!seen[v] && this.level[v] > 0) {
            seen[v] = true; this._bump(v);
            if (this.level[v] >= curLevel) pathCount++;
            else learnt.push(q);
          }
        }
        while (!seen[Math.abs(this.trail[index])]) index--;
        p = this.trail[index];
        seen[Math.abs(p)] = false;
        index--; pathCount--;
        if (pathCount <= 0) break;
        lits = this.clauses[this.reason[Math.abs(p)]];
      }
      learnt[0] = -p;
      const min = this._minimize(learnt);
      let bt;
      if (min.length === 1) bt = 0;
      else {
        let maxI = 1;
        for (let i = 2; i < min.length; i++)
          if (this.level[Math.abs(min[i])] > this.level[Math.abs(min[maxI])]) maxI = i;
        const t = min[1]; min[1] = min[maxI]; min[maxI] = t;
        bt = this.level[Math.abs(min[1])];
      }
      return { learnt: min, bt: bt, uip: -learnt[0] };
    }

    _minimize(learnt) {
      const mark = new Set(learnt.map(l => Math.abs(l)));
      const out = [learnt[0]];
      for (let i = 1; i < learnt.length; i++) {
        const l = learnt[i];
        if (this.reason[Math.abs(l)] === null) out.push(l);
        else if (!this._redundant(l, mark, 0)) out.push(l);
      }
      return out;
    }
    _redundant(lit, mark, depth) {
      const r = this.reason[Math.abs(lit)];
      if (r === null) return false;
      if (depth > 200) return false;
      for (const q of this.clauses[r]) {
        if (Math.abs(q) === Math.abs(lit)) continue;
        const v = Math.abs(q);
        if (this.level[v] === 0) continue;
        if (mark.has(v)) continue;
        if (this.reason[v] === null) return false;
        if (!this._redundant(q, mark, depth + 1)) return false;
      }
      return true;
    }

    solve(maxConflicts) {
      if (!this.ok) return "UNSAT";
      if (this.heap.size === 0) {
        for (let v = 1; v <= this.numVars; v++) this.heap.push({ act: this.activity[v], v: v });
      }
      if (this.propagate() !== null) { this.ok = false; return "UNSAT"; }
      let restartNo = 0;
      let untilRestart = this.restartBase * luby(restartNo + 1);
      let thisRun = 0;
      while (true) {
        const confl = this.propagate();
        if (confl !== null) {
          this.stats.conflicts++; thisRun++;
          if (maxConflicts && this.stats.conflicts > maxConflicts) return "UNKNOWN";
          this._emit({ t: "conflict", clause: this.clauses[confl].slice(),
                       level: this.decisionLevel(), graph: this._conflictGraph(confl) });
          if (this.decisionLevel() === 0) return "UNSAT";
          const { learnt, bt, uip } = this.analyze(confl);
          this.stats.learned++;
          this._emit({ t: "learn", clause: learnt.slice(), bt: bt, uip: uip });
          this._backtrackTo(bt);
          let acref = null;
          if (learnt.length === 1) this._enqueue(learnt[0], null);
          else { acref = this._addLearnt(learnt); this._enqueue(learnt[0], acref); }
          this._emit({ t: "backjump", to: bt, lit: learnt[0] });
          this._decay();
        } else {
          if (this.decisionLevel() > this.stats.maxLevel) this.stats.maxLevel = this.decisionLevel();
          if (thisRun >= untilRestart) {
            this.stats.restarts++; restartNo++; thisRun = 0;
            untilRestart = this.restartBase * luby(restartNo + 1);
            this._backtrackTo(0);
            this._emit({ t: "restart" });
            continue;
          }
          const v = this._pickBranchVar();
          if (v === null) return "SAT";
          this.stats.decisions++;
          this._newDecisionLevel();
          const lit = this.phase[v] ? v : -v;
          this._enqueue(lit, null);
          this._emit({ t: "decide", lit: lit, level: this.decisionLevel() });
        }
      }
    }

    // snapshot the implication-graph cone feeding a conflict (for the viz)
    _conflictGraph(confl) {
      const nodes = {}, edges = [];
      const addNode = (lit) => {
        const v = Math.abs(lit);
        if (!(v in nodes)) nodes[v] = { lit: this.val[v] ? v : -v, level: this.level[v],
                                        decision: this.reason[v] === null };
      };
      const queue = [];
      for (const l of this.clauses[confl]) { addNode(l); queue.push(Math.abs(l)); edges.push({ from: -l, to: "K" }); }
      const visited = new Set();
      while (queue.length) {
        const v = queue.shift();
        if (visited.has(v)) continue; visited.add(v);
        const r = this.reason[v];
        if (r === null) continue;
        const implied = this.clauses[r][0];
        for (const q of this.clauses[r]) {
          if (Math.abs(q) === v) continue;
          addNode(q); edges.push({ from: -q, to: implied }); queue.push(Math.abs(q));
        }
      }
      return { nodes: Object.values(nodes), edges: edges, conflictClause: this.clauses[confl].slice() };
    }

    model() {
      const m = {};
      for (let v = 1; v <= this.numVars; v++) m[v] = this.val[v] === null ? false : this.val[v];
      return m;
    }
  }

  function isSatisfied(numVars, clauses, model) {
    for (const cl of clauses) {
      let ok = false;
      for (const l of cl) { if (model[Math.abs(l)] === (l > 0)) { ok = true; break; } }
      if (!ok) return false;
    }
    return true;
  }

  function solveCNF(numVars, clauses, opts) {
    const s = new Solver(numVars, opts);
    for (const cl of clauses) if (!s.addClause(cl)) break;
    const res = s.solve(opts && opts.maxConflicts);
    return { result: res, model: res === "SAT" ? s.model() : null, solver: s };
  }

  const api = { Solver, solveCNF, isSatisfied, luby };
  if (typeof module !== "undefined" && module.exports) module.exports = api;
  root.Conflux = api;
})(typeof window !== "undefined" ? window : globalThis);
